pallets/treasury/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// It transfers balances from treasury to each of beneficiaries and the specific amount
+        /// for each of them.
+        ///
+        /// # Error
+        /// * `BadOrigin`: Only root can execute transaction.
+        /// * `InsufficientBalance`: If treasury balances is not enough to cover all beneficiaries.
+        pub fn disbursement(origin, beneficiaries: Vec< Beneficiary<T::Balance>>) -> DispatchResult
+        {
+            ensure_root(origin)?;
+
+            // Ensure treasury has enough balance.
+            let total_amount = beneficiaries.iter().fold( 0.into(), |acc,b| b.amount.saturating_add(acc));
+            ensure!(
+                Self::balance() >= total_amount,
+                Error::<T>::InsufficientBalance);
+
+            beneficiaries.into_iter().for_each( |b| {
+                Self::unsafe_disbursement(b.id, b.amount);
+                Self::deposit_event(RawEvent::TreasuryDisbursement(b.id, b.amount));
+            });
+            Ok(())
+        }
+
+        /// It transfers the specific `amount` from `origin` account into treasury.
+        ///
+        /// Only accounts which are associated to an identity can make a donation to treasury.
+        pub fn reimbursement(origin, amount: T::Balance) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let sender_key = AccountKey::try_from(sender.encode())?;
+            let _did = Context::current_identity_or::<Identity<T>>(&sender_key)?;
+
+            let _ = balances::Module::<T>::withdraw(
+                &sender,
+                amount,
+                WithdrawReason::Transfer.into(),
+                ExistenceRequirement::AllowDeath,
+            )?;
+
+            Self::unsafe_reimbursement(amount);
+            Self::deposit_event(RawEvent::TreasuryReimbursement(amount));
+            Ok(())
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    pub fn unsafe_disbursement(target: IdentityId, amount: T::Balance) {
+        let new_treasury_balance = Self::balance() - amount;
+        <Balance<T>>::put(new_treasury_balance);
+
+        balances::Module::<T>::unsafe_top_up_identity_balance(&target, amount);
+    }
+
+    fn unsafe_reimbursement(amount: T::Balance) {
+        let new_balance = Self::balance().saturating_add(amount);
+        <Balance<T>>::put(new_balance);
+    }
+}
+
+impl<T: Trait> TreasuryTrait<T::Balance> for Module<T> {
+    #[inline]
+    fn disbursement(target: IdentityId, amount: T::Balance) {
+        Self::unsafe_disbursement(target, amount);
+    }
+
+    #[inline]
+    fn balance() -> T::Balance {
+        Self::balance()
+    }
+}
+
+impl<T: Trait> OnUnbalanced<NegativeImbalance<T>> for Module<T> {
+    fn on_nonzero_unbalanced(amount: NegativeImbalance<T>) {
+        let abs_amount = amount.peek();
+        Self::unsafe_reimbursement(abs_amount);
```

Comment:
This will trigger `unsafe_reimbursement` twice. Once when you manually call it in `reimbursement` fn and then again when this imbalance is dropped.